diff --git a/lib/Flux/Athena.php b/lib/Flux/Athena.php
index e9d9ad5..b852995 100644
--- a/lib/Flux/Athena.php
+++ b/lib/Flux/Athena.php
@@ -324,7 +324,7 @@ class Flux_Athena {
 		$xferTable    = Flux::config('FluxTables.CreditTransferTable');
 		
 		// Get balance of sender.
-		$sql = "SELECT balance FROM {$this->loginDatabase}.$creditsTable WHERE account_id = ? LIMIT 1";
+		$sql = "SELECT value FROM {$this->charMapDatabase}.`global_reg_value` where `account_id` = ? and `str` = '#CASHPOINTS' LIMIT 1";
 		$sth = $this->connection->getStatement($sql);
 		
 		if (!$sth->execute(array($fromAccountID))) {
@@ -332,7 +332,7 @@ class Flux_Athena {
 			return false;
 		}
 		
-		if ($sth->fetch()->balance < $credits) {
+		if ($sth->fetch()->value < $credits) {
 			// Insufficient balance.
 			return -2;
 		}
diff --git a/lib/Flux/ItemShop/Cart.php b/lib/Flux/ItemShop/Cart.php
index e62270b..2df1fc4 100644
--- a/lib/Flux/ItemShop/Cart.php
+++ b/lib/Flux/ItemShop/Cart.php
@@ -114,7 +114,7 @@ class Flux_ItemShop_Cart {
 	public function hasFunds()
 	{
 		$this->requiresAccount();
-		$creditsAvailable = $this->account->balance;
+		$creditsAvailable = $this->account->value;
 		$creditsNeeded    = $this->getTotal();
 		
 		if ($creditsAvailable < $creditsNeeded) {
diff --git a/lib/Flux/LoginServer.php b/lib/Flux/LoginServer.php
index 7d35e63..e9e7963 100644
--- a/lib/Flux/LoginServer.php
+++ b/lib/Flux/LoginServer.php
@@ -359,7 +359,7 @@ class Flux_LoginServer extends Flux_BaseServer {
 	{
 		$creditsTable = Flux::config('FluxTables.CreditsTable');
 		
-		$sql = "SELECT COUNT(account_id) AS hasRecord FROM {$this->loginDatabase}.$creditsTable WHERE account_id = ?";
+		$sql = "SELECT COUNT(account_id) AS hasRecord FROM {$this->charMapDatabase}.`global_reg_value` where `account_id` = ? and `str` = '#CASHPOINTS'";
 		$sth = $this->connection->getStatement($sql);
 		
 		$sth->execute(array($accountID));
@@ -387,39 +387,39 @@ class Flux_LoginServer extends Flux_BaseServer {
 		$creditsTable = Flux::config('FluxTables.CreditsTable');
 		
 		if (!$this->hasCreditsRecord($targetAccountID)) {
-			$fields = 'account_id, balance';
-			$values = '?, ?';
+			$fields = 'char_id, str, value, type, account_id';
+			$values = '?, ?, ?, ?, ?';
 			
-			if (!is_null($donationAmount)) {
+			/*if (!is_null($donationAmount)) {
 				$fields .= ', last_donation_date, last_donation_amount';
 				$values .= ', NOW(), ?';
-			}
+			}*/
 			
-			$sql  = "INSERT INTO {$this->loginDatabase}.$creditsTable ($fields) VALUES ($values)";
+			$sql  = "INSERT INTO {$this->charMapDatabase}.`global_reg_value` ($fields) VALUES ($values)";
 			$sth  = $this->connection->getStatement($sql);
-			$vals = array($targetAccountID, $credits);
+			$vals = array(0, '#CASHPOINTS', $credits, 2, $targetAccountID);
 			
-			if (!is_null($donationAmount)) {
+			/*if (!is_null($donationAmount)) {
 				$vals[] = $donationAmount;
-			}
+			}*/
 			
 			return $sth->execute($vals);
 		}
 		else {
 			$vals = array();
-			$sql  = "UPDATE {$this->loginDatabase}.$creditsTable SET balance = balance + ? ";
+			$sql  = "UPDATE {$this->charMapDatabase}.`global_reg_value` SET value = value + ? ";
 
-			if (!is_null($donationAmount)) {
+			/*if (!is_null($donationAmount)) {
 				$sql .= ", last_donation_date = NOW(), last_donation_amount = ? ";
-			}
+			}*/
 			
 			$vals[] = $credits;
-			if (!is_null($donationAmount)) {
+			/*if (!is_null($donationAmount)) {
 				$vals[] = $donationAmount;
-			}
+			}*/
 			$vals[] = $targetAccountID;
 			
-			$sql .= "WHERE account_id = ?";
+			$sql .= "WHERE account_id = ? and `str` = '#CASHPOINTS'";
 			$sth  = $this->connection->getStatement($sql);
 			
 			return $sth->execute($vals);
diff --git a/lib/Flux/PaymentNotifyRequest.php b/lib/Flux/PaymentNotifyRequest.php
index a113c0d..8914f35 100644
--- a/lib/Flux/PaymentNotifyRequest.php
+++ b/lib/Flux/PaymentNotifyRequest.php
@@ -215,12 +215,12 @@ class Flux_PaymentNotifyRequest {
 								$credits = floor($amount / $rate);
 
 								if ($trusted) {
-									$sql = "SELECT * FROM {$servGroup->loginDatabase}.{$this->creditsTable} WHERE account_id = ?";
+									$sql = "SELECT * FROM {$this->charMapDatabase}.`global_reg_value` WHERE account_id = ? and `str` = '#CASHPOINTS'";
 									$sth = $servGroup->connection->getStatement($sql);
 									$sth->execute(array($accountID));
 									$acc = $sth->fetch();
 
-									$this->logPayPal('Updating account credit balance from %s to %s', (int)$acc->balance, $acc->balance + $credits);
+									$this->logPayPal('Updating account credit balance from %s to %s', (int)$acc->value, $acc->value + $credits);
 									$res = $servGroup->loginServer->depositCredits($accountID, $credits, $amount);
 
 									if ($res) {
diff --git a/lib/Flux/SessionData.php b/lib/Flux/SessionData.php
index 02635f3..a526516 100644
--- a/lib/Flux/SessionData.php
+++ b/lib/Flux/SessionData.php
@@ -288,10 +288,10 @@ class Flux_SessionData {
 		}
 		
 		$creditsTable  = Flux::config('FluxTables.CreditsTable');
-		$creditColumns = 'credits.balance, credits.last_donation_date, credits.last_donation_amount';
+		$creditColumns = 'credits.value';
 		
 		$sql  = "SELECT login.*, {$creditColumns} FROM {$loginAthenaGroup->loginDatabase}.login ";
-		$sql .= "LEFT OUTER JOIN {$loginAthenaGroup->loginDatabase}.{$creditsTable} AS credits ON login.account_id = credits.account_id ";
+		$sql .= "LEFT OUTER JOIN {$this->charMapDatabase}.`global_reg_value` AS credits ON login.account_id = credits.account_id and credits.`str` = '#CASHPOINTS'";
 		$sql .= "WHERE login.sex != 'S' AND login.group_id >= 0 AND login.userid = ? LIMIT 1";
 		$smt  = $loginAthenaGroup->connection->getStatement($sql);
 		$res  = $smt->execute(array($username));
@@ -348,10 +348,10 @@ class Flux_SessionData {
 	private function getAccount(Flux_LoginAthenaGroup $loginAthenaGroup, $username)
 	{
 		$creditsTable  = Flux::config('FluxTables.CreditsTable');
-		$creditColumns = 'credits.balance, credits.last_donation_date, credits.last_donation_amount';
+		$creditColumns = 'credits.value';
 		
 		$sql  = "SELECT login.*, {$creditColumns} FROM {$loginAthenaGroup->loginDatabase}.login ";
-		$sql .= "LEFT OUTER JOIN {$loginAthenaGroup->loginDatabase}.{$creditsTable} AS credits ON login.account_id = credits.account_id ";
+		$sql .= "LEFT OUTER JOIN {$this->charMapDatabase}.`global_reg_value` AS credits ON login.account_id = credits.account_id and credits.`str` = '#CASHPOINTS' ";
 		$sql .= "WHERE login.sex != 'S' AND login.group_id >= 0 AND login.userid = ? LIMIT 1";
 		$smt  = $loginAthenaGroup->connection->getStatement($sql);
 		$res  = $smt->execute(array($username));
diff --git a/modules/account/changesex.php b/modules/account/changesex.php
index 553939c..3c75aa5 100644
--- a/modules/account/changesex.php
+++ b/modules/account/changesex.php
@@ -8,7 +8,7 @@ $title = Flux::message('GenderChangeTitle');
 $cost    = +(int)Flux::config('ChargeGenderChange');
 $badJobs = Flux::config('GenderLinkedJobClasses')->toArray();
 
-if ($cost && $session->account->balance < $cost && !$auth->allowedToAvoidSexChangeCost) {
+if ($cost && $session->account->value < $cost && !$auth->allowedToAvoidSexChangeCost) {
 	$hasNecessaryFunds = false;
 }
 else {
diff --git a/modules/account/edit.php b/modules/account/edit.php
index 54e582b..5330db8 100644
--- a/modules/account/edit.php
+++ b/modules/account/edit.php
@@ -8,10 +8,10 @@ $title = Flux::message('AccountEditTitle');
 $accountID = $params->get('id');
 
 $creditsTable  = Flux::config('FluxTables.CreditsTable');
-$creditColumns = 'credits.balance, credits.last_donation_date, credits.last_donation_amount';
+$creditColumns = 'credits.value';
 
 $sql  = "SELECT login.*, {$creditColumns} FROM {$server->loginDatabase}.login ";
-$sql .= "LEFT OUTER JOIN {$creditsTable} AS credits ON login.account_id = credits.account_id ";
+$sql .= "LEFT OUTER JOIN {$this->charMapDatabase}.`global_reg_value` AS credits ON login.account_id = credits.account_id and credits.`str` = '#CASHPOINTS' ";
 $sql .= "WHERE login.sex != 'S' AND login.group_id >= 0 AND login.account_id = ? LIMIT 1";
 $sth  = $server->connection->getStatement($sql);
 $sth->execute(array($accountID));
@@ -61,7 +61,7 @@ if ($account) {
 		elseif (!in_array($gender, array('M', 'F'))) {
 			$errorMessage = Flux::message('InvalidGender');
 		}
-		elseif ($account->balance != $balance && !$auth->allowedToEditAccountBalance) {
+		elseif ($account->value != $balance && !$auth->allowedToEditAccountBalance) {
 			$errorMessage = Flux::message('CannotModifyBalance');
 		}
 		elseif ($birthdate && !preg_match('/^\d{4}-\d{2}-\d{2}$/', $birthdate)) {
@@ -95,7 +95,7 @@ if ($account) {
 			$sth->execute($bind);
 
 			if ($auth->allowedToEditAccountBalance) {
-				$deposit = $balance - $account->balance;
+				$deposit = $balance - $account->value;
 				$session->loginServer->depositCredits($account->account_id, $deposit);
 			}
 			
diff --git a/modules/account/index.php b/modules/account/index.php
index 6d33e83..26a79cd 100644
--- a/modules/account/index.php
+++ b/modules/account/index.php
@@ -17,12 +17,12 @@ $searchPassword = (($useMD5 && $searchMD5) || !$useMD5) && $auth->allowedToSeeAc
 $showPassword   = !$useMD5 && $auth->allowedToSeeAccountPassword;
 $bind           = array();
 $creditsTable   = Flux::config('FluxTables.CreditsTable');
-$creditColumns  = 'credits.balance, credits.last_donation_date, credits.last_donation_amount';
+$creditColumns  = 'credits.value';
 $accountTable   = Flux::config('FluxTables.AccountCreateTable');
 $accountColumns = 'createlog.reg_date';
 $createTable    = Flux::config('FluxTables.AccountCreateTable');
 $createColumns  = 'created.confirmed, created.confirm_code, created.reg_date';
-$sqlpartial     = "LEFT OUTER JOIN {$server->loginDatabase}.{$creditsTable} AS credits ON login.account_id = credits.account_id ";
+$sqlpartial     = "LEFT OUTER JOIN {$this->charMapDatabase}.`global_reg_value` AS credits ON login.account_id = credits.account_id and credits.`str` = '#CASHPOINTS' ";
 $sqlpartial    .= "LEFT OUTER JOIN {$server->loginDatabase}.{$accountTable} AS createlog ON login.account_id = createlog.account_id ";
 $sqlpartial    .= "LEFT OUTER JOIN {$server->loginDatabase}.{$createTable} AS created ON login.account_id = created.account_id ";
 $sqlpartial    .= "WHERE login.sex != 'S' AND login.group_id >= 0 ";
@@ -44,7 +44,7 @@ else {
 	$accountGroupIdOp = $params->get('account_group_id_op');
 	$accountGroupID   = $params->get('account_group_id');
 	$balanceOp        = $params->get('balance_op');
-	$balance          = $params->get('balance');
+	$balance          = $params->get('value');
 	$loginCountOp     = $params->get('logincount_op');
 	$loginCount       = $params->get('logincount');
 	$birthdateA       = $params->get('birthdate_after_date');
@@ -111,10 +111,10 @@ else {
 	if (in_array($balanceOp, $opValues) && trim($balance) != '') {
 		$op  = $opMapping[$balanceOp];
 		if ($op == '=' && $balance === '0') {
-			$sqlpartial .= "AND (credits.balance IS NULL OR credits.balance = 0) ";
+			$sqlpartial .= "AND (credits.value IS NULL OR credits.value = 0) ";
 		}
 		else {
-			$sqlpartial .= "AND credits.balance $op ? ";
+			$sqlpartial .= "AND credits.value $op ? ";
 			$bind[]      = $balance;
 		}
 	}
@@ -153,7 +153,7 @@ $sth->execute($bind);
 $paginator = $this->getPaginator($sth->fetch()->total);
 $paginator->setSortableColumns(array(
 	'login.account_id' => 'asc', 'login.userid', 'login.user_pass',
-	'login.sex', 'group_id', 'state', 'balance',
+	'login.sex', 'group_id', 'state', 'value',
 	'login.email', 'logincount', 'lastlogin', 'last_ip',
 	'reg_date'
 ));
diff --git a/modules/account/transfer.php b/modules/account/transfer.php
index b194408..9a0ee0e 100644
--- a/modules/account/transfer.php
+++ b/modules/account/transfer.php
@@ -6,7 +6,7 @@ $this->loginRequired();
 $title = Flux::message('TransferTitle');
 
 if (count($_POST)) {
-	if ($session->account->balance) {
+	if ($session->account->value) {
 		$credits  = (int)$params->get('credits'); 
 		$charName = trim($params->get('char_name'));
 		
diff --git a/modules/account/view.php b/modules/account/view.php
index 73733af..823970c 100644
--- a/modules/account/view.php
+++ b/modules/account/view.php
@@ -16,7 +16,7 @@ $tableName = "{$server->charMapDatabase}.items";
 $tempTable = new Flux_TemporaryTable($server->connection, $tableName, $fromTables);
 
 $creditsTable  = Flux::config('FluxTables.CreditsTable');
-$creditColumns = 'credits.balance, credits.last_donation_date, credits.last_donation_amount';
+$creditColumns = 'credits.value';
 $createTable   = Flux::config('FluxTables.AccountCreateTable');
 $createColumns = 'created.confirmed, created.confirm_code, created.reg_date';
 $isMine        = false;
@@ -36,7 +36,7 @@ if (!$isMine) {
 	}
 	
 	$sql  = "SELECT login.*, {$creditColumns}, {$createColumns} FROM {$server->loginDatabase}.login ";
-	$sql .= "LEFT OUTER JOIN {$server->loginDatabase}.{$creditsTable} AS credits ON login.account_id = credits.account_id ";
+	$sql .= "LEFT OUTER JOIN {$this->charMapDatabase}.`global_reg_value` AS credits ON login.account_id = credits.account_id and credits.`str` = '#CASHPOINTS' ";
 	$sql .= "LEFT OUTER JOIN {$server->loginDatabase}.{$createTable} AS created ON login.account_id = created.account_id ";
 	$sql .= "WHERE login.sex != 'S' AND login.group_id >= 0 AND login.account_id = ? LIMIT 1";
 	$sth  = $server->connection->getStatement($sql);
diff --git a/modules/purchase/checkout.php b/modules/purchase/checkout.php
index cdc3fe4..84e2917 100644
--- a/modules/purchase/checkout.php
+++ b/modules/purchase/checkout.php
@@ -26,7 +26,7 @@ if (count($_POST) && $params->get('process')) {
 	$sql .= "VALUES (?, ?, ?, ?, NULL, 0, NULL, NOW(), ?, ?)";
 	$sth  = $server->connection->getStatement($sql);
 	
-	$balance = $session->account->balance;
+	$balance = $session->account->value;
 	
 	foreach ($items as $item) {
 		$creditsAfter = $balance - $item->shop_item_cost;
diff --git a/themes/default/account/changesex.php b/themes/default/account/changesex.php
index 4ca963b..ab3e6c3 100644
--- a/themes/default/account/changesex.php
+++ b/themes/default/account/changesex.php
@@ -3,7 +3,7 @@
 <?php if ($cost): ?>
 <p>
 	<?php printf(Flux::message('GenderChangeCost'), '<span class="remaining-balance">'.number_format((int)$cost).'</span>') ?>
-	<?php printf(Flux::message('GenderChangeBalance'), '<span class="remaining-balance">'.number_format((int)$session->account->balance).'</span>') ?>
+	<?php printf(Flux::message('GenderChangeBalance'), '<span class="remaining-balance">'.number_format((int)$session->account->value).'</span>') ?>
 </p>
 <?php if (!$hasNecessaryFunds): ?>
 <p><?php echo htmlspecialchars(Flux::message('GenderChangeNoFunds')) ?></p>
diff --git a/themes/default/account/edit.php b/themes/default/account/edit.php
index b21b7cc..b2467a2 100644
--- a/themes/default/account/edit.php
+++ b/themes/default/account/edit.php
@@ -53,10 +53,10 @@
 				<td><input type="text" name="logincount" id="logincount" value="<?php echo (int)$account->logincount ?>" /></td>
 				<?php if ($auth->allowedToEditAccountBalance): ?>
 					<th><label for="balance"><?php echo htmlspecialchars(Flux::message('CreditBalanceLabel')) ?></label></th>
-					<td><input type="text" name="balance" id="balance" value="<?php echo (int)$account->balance ?>" /></td>
+					<td><input type="text" name="balance" id="balance" value="<?php echo (int)$account->value ?>" /></td>
 				<?php else: ?>
 					<th><?php echo htmlspecialchars(Flux::message('CreditBalanceLabel')) ?></th>
-					<td><?php echo number_format((int)$account->balance) ?></td>
+					<td><?php echo number_format((int)$account->value) ?></td>
 				<?php endif ?>
 			</tr>
 			<tr>
diff --git a/themes/default/account/index.php b/themes/default/account/index.php
index f4dea70..4270053 100644
--- a/themes/default/account/index.php
+++ b/themes/default/account/index.php
@@ -134,7 +134,7 @@
 				<span class="account-state state-unknown"><?php echo htmlspecialchars(Flux::message('UnknownLabel')) ?></span>
 			<?php endif ?>
 		</td>
-		<td><?php echo number_format((int)$account->balance) ?></td>
+		<td><?php echo number_format((int)$account->value) ?></td>
 		<td>
 			<?php if ($account->email): ?>
 				<?php echo $this->linkToAccountSearch(array('email' => $account->email), $account->email) ?>
diff --git a/themes/default/account/transfer.php b/themes/default/account/transfer.php
index b646dfd..632a826 100644
--- a/themes/default/account/transfer.php
+++ b/themes/default/account/transfer.php
@@ -3,9 +3,9 @@
 <?php if (!empty($errorMessage)): ?>
 	<p class="red"><?php echo htmlspecialchars($errorMessage) ?></p>
 <?php endif ?>
-<?php if ($session->account->balance): ?>
+<?php if ($session->account->value): ?>
 <h3><?php printf(htmlspecialchars(Flux::message('TransferSubHeading')), $server->serverName) ?></h3>
-<p><?php printf(Flux::message('TransferInfo'), '<span class="remaining-balance">'.number_format($session->account->balance).'</span>') ?></p>
+<p><?php printf(Flux::message('TransferInfo'), '<span class="remaining-balance">'.number_format($session->account->value).'</span>') ?></p>
 <p><?php echo htmlspecialchars(Flux::message('TransferInfo2')) ?></p>
 <form action="<?php echo $this->url ?>" method="post" class="generic-form">
 	<?php echo $this->moduleActionFormInputs('account', 'transfer') ?>
diff --git a/themes/default/account/view.php b/themes/default/account/view.php
index fb9df90..65bd834 100644
--- a/themes/default/account/view.php
+++ b/themes/default/account/view.php
@@ -64,7 +64,7 @@
 		<td><?php echo number_format((int)$account->logincount) ?></td>
 		<th><?php echo htmlspecialchars(Flux::message('CreditBalanceLabel')) ?></th>
 		<td>
-			<?php echo number_format((int)$account->balance) ?>
+			<?php echo number_format((int)$account->value) ?>
 			<?php if ($auth->allowedToDonate && $isMine): ?>
 				<a href="<?php echo $this->url('donate') ?>"><?php echo htmlspecialchars(Flux::message('AccountViewDonateLink')) ?></a>
 			<?php endif ?>
diff --git a/themes/default/main/balance.php b/themes/default/main/balance.php
index 7370870..ff6b932 100644
--- a/themes/default/main/balance.php
+++ b/themes/default/main/balance.php
@@ -1,5 +1,5 @@
 <?php if (!defined('FLUX_ROOT')) exit; ?>
 <div class="credit-balance">
 	<span class="balance-text">Donation Credits</span>
-	<span class="balance-amount"><?php echo number_format((int)$session->account->balance) ?></span>
+	<span class="balance-amount"><?php echo number_format((int)$session->account->value) ?></span>
 </div>
\ No newline at end of file
diff --git a/themes/default/purchase/checkout.php b/themes/default/purchase/checkout.php
index 48ac63a..62ebeac 100644
--- a/themes/default/purchase/checkout.php
+++ b/themes/default/purchase/checkout.php
@@ -4,7 +4,7 @@
 
 <h3>Purchase Information</h3>
 <p class="cart-total-text">Your current subtotal is <span class="cart-sub-total"><?php echo number_format($total=$server->cart->getTotal()) ?></span> credit(s).</p>
-<p class="checkout-info-text">Your remaining balance after this purchase will be <span class="remaining-balance"><?php echo number_format($session->account->balance - $total) ?></span> credit(s).</p>
+<p class="checkout-info-text">Your remaining balance after this purchase will be <span class="remaining-balance"><?php echo number_format($session->account->value - $total) ?></span> credit(s).</p>
 <p>After reviewing the below item information, you can proceed with your checkout by clicking the “Purchase Items” button.</p>
 <p class="important">Note: These items are for redemption on the <span class="server-name"><?php echo htmlspecialchars($server->serverName) ?></span> server ONLY.</p>
 <p>
